apps/prairielearn/src/tests/groupExamRolePermissions.test.ts

```diff
@@ -0,0 +1,583 @@
+    method: 'POST',
+    body: new URLSearchParams({
+      __action: 'create_group',
+      __csrf_token: csrfToken,
+      groupName: groupName,
+    }),
+  });
+  assert.isOk(res.ok);
+  const $ = cheerio.load(await res.text());
+  return $;
+}
+
+/**
+ * Joins a group in an assessment using the provided join code as the user with the given CSRF token
+ */
+async function joinGroup(
+  assessmentUrl: string,
+  joinCode: string,
+  csrfToken: string,
+): Promise<cheerio.CheerioAPI> {
+  const res = await fetch(assessmentUrl, {
+    method: 'POST',
+    body: new URLSearchParams({
+      __action: 'join_group',
+      __csrf_token: csrfToken,
+      join_code: joinCode,
+    }),
+  });
+  assert.isOk(res.ok);
+  const $ = cheerio.load(await res.text());
+  return $;
+}
+
+/**
+ * Sends and verifies a group roles update request using current user.
+ * Updates element list to check that group role select table is changed correctly.
+ */
+async function updateGroupRoles(
+  roleUpdates: any[],
+  groupRoles: any[],
+  studentUsers: StudentUser[],
+  csrfToken: string,
+  assessmentUrl: string,
+  $: cheerio.CheerioAPI,
+): Promise<cheerio.CheerioAPI> {
+  // Uncheck all of the inputs
+  const roleIds = groupRoles.map((role) => role.id);
+  const userIds = studentUsers.map((user) => user.user_id);
+  for (const roleId of roleIds) {
+    for (const userId of userIds) {
+      const elementId = `#user_role_${roleId}-${userId}`;
+      $('#role-select-form').find(elementId).removeAttr('checked');
+    }
+  }
+
+  let checkedBoxes = $('#role-select-form').find('tr').find('input:checked');
+  assert.lengthOf(checkedBoxes, 0, 'all checkboxes in role select form must be unchecked');
+
+  // Mark the checkboxes as checked
+  roleUpdates.forEach(({ roleId, groupUserId }) => {
+    $(`#user_role_${roleId}-${groupUserId}`).attr('checked', '');
+  });
+  checkedBoxes = $('#role-select-form').find('tr').find('input:checked');
+  assert.lengthOf(
+    checkedBoxes,
+    roleUpdates.length,
+    'all checkboxes in role select form must be checked',
+  );
+
+  // Grab IDs of checkboxes to construct update request
+  const checkedElementIds = {};
+  for (let i = 0; i < checkedBoxes.length; i++) {
+    checkedElementIds[checkedBoxes[i.toString()].attribs.id] = 'on';
+  }
+  const res = await fetch(assessmentUrl, {
+    method: 'POST',
+    body: new URLSearchParams({
+      __action: 'update_group_roles',
+      __csrf_token: csrfToken,
+      ...checkedElementIds,
+    }),
+  });
+  assert.isOk(res.ok, 'updating group roles should be successful');
+  return cheerio.load(await res.text());
+}
+
+async function getQuestionUrl(
+  courseInstanceUrl: string,
+  assessmentInstanceId: string,
+  questionId: string,
+): Promise<string> {
+  const result = await queryValidatedOneRow(
+    sql.select_instance_questions,
+    {
+      assessment_instance_id: assessmentInstanceId,
+      question_id: questionId,
+    },
+    QuestionIdSchema,
```

Comment:
Nit: why not use `IdSchema` directly here?

Reply:
Again, I didn't rewrite the tests, so a few minor issues remained. This code (and one other call) was still using `queryValidatedOneRow`, so the object was returned with an id column instead of the single value. I replaced this in ba2016c41.